x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+     * Writes the authentication to the {@link ThreadContext} and then calls the listener if
+     * successful
+     */
+    void writeAuthToContext(Authenticator.Context context, Authentication authentication, ActionListener<Authentication> listener) {
+        try {
+            authenticationSerializer.writeToContext(authentication, context.getThreadContext());
+            context.getRequest().authenticationSuccess(authentication);
+            // Header for operator privileges will only be written if authentication actually happens,
+            // i.e. not read from either header or transient header
+            operatorPrivilegesService.maybeMarkOperatorUser(authentication, context.getThreadContext());
+        } catch (Exception e) {
+            logger.debug(new ParameterizedMessage("Failed to store authentication [{}] for request [{}]",
+                authentication,
+                context.getRequest()), e);
+            final ElasticsearchSecurityException ese = context.getRequest().exceptionProcessingRequest(e, context.getAuthenticationToken());
+            addMetadata(context, ese);
+            listener.onFailure(ese);
+            return;
+        }
+        logger.trace("Established authentication [{}] for request [{}]", authentication, context.getRequest());
+        listener.onResponse(authentication);
+    }
+
+    private void addMetadata(Authenticator.Context context, ElasticsearchSecurityException ese) {
+        if (false == context.getUnsuccessfulMessages().isEmpty()) {
+            ese.addMetadata("es.additional_unsuccessful_credentials", context.getUnsuccessfulMessages());
+        }
+    }
+
+    /**
+     * Determines whether to support anonymous access for the current request. Returns {@code true} if all of the following are true
+     * <ul>
+     *     <li>The service has anonymous authentication enabled (see {@link #isAnonymousUserEnabled})</li>
+     *     <li>Anonymous access is accepted for this request ({@code allowAnonymousOnThisRequest} parameter)
+     *     <li>The {@link ThreadContext} does not provide API Key or Bearer Token credentials. If these are present, we
+     *     treat the request as though it attempted to authenticate (even if that failed), and will not fall back to anonymous.</li>
+     * </ul>
+     */
+    private boolean shouldFallbackToAnonymous(Authenticator.Context context) {
+        if (isAnonymousUserEnabled == false) {
+            return false;
+        }
+        if (context.isAllowAnonymous() == false) {
+            return false;
+        }
+        String header = context.getThreadContext().getHeader("Authorization");
+        if (Strings.hasText(header) &&
+            ((header.regionMatches(true, 0, "Bearer ", 0, "Bearer ".length()) && header.length() > "Bearer ".length()) ||
+                (header.regionMatches(true, 0, "ApiKey ", 0, "ApiKey ".length()) && header.length() > "ApiKey ".length()))) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
This existing method basically means anonymous is not allowed if any credentials can be found in the context. But this is not the case for fallback user, which is weird. I think both of them, fallback and anonymous, should be handled the same way.